bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniFormDetailsServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+
+        List<Encounter> encounters = getEncounters(limitedVisits);
+
+        if (isNotEmpty(encounters) && isNotEmpty(limitedVisits)) {
+            return getFormDetails(patient, encounters, formType);
+        }
+        return Collections.emptyList();
+    }
+
+    private Patient getPatient(String patientUuid) {
+        Patient patient = patientService.getPatientByUuid(patientUuid);
+        if (patient == null) {
+            throw new InvalidParameterException("Patient does not exist");
+        }
+        return patient;
+    }
+
+    private List<Encounter> getEncounters(List<Visit> visits) {
+        EncounterSearchCriteria encounterSearchCriteria = new EncounterSearchCriteriaBuilder()
+                .setVisits(visits).createEncounterSearchCriteria();
+        return encounterService.getEncounters(encounterSearchCriteria);
+    }
+
+    private Collection<FormDetails> getFormDetails(Patient patient, List<Encounter> encounters, String formType) {
+        Collection<FormDetails> formDetails = new ArrayList<>();
+        List<Obs> observations = obsService.getObservations(singletonList(patient.getPerson()), encounters,
+                null, null, null, null, null, null, null, null, null, false);
+        if (FormType.FORM_BUILDER_FORMS.get().equals(formType) || StringUtils.isBlank(formType)) {
+            formDetails = createFormDetails(ObsUtil.filterFormBuilderObs(observations), FormType.FORM_BUILDER_FORMS);
+        }
+        return formDetails;
+    }
+
+    @Override
+    public Collection<FormDetails> getFormDetails(String patientUuid, String formType, String visitUuid,
+                                                  String patientProgramUuid) {
+        Patient patient = getPatient(patientUuid);
+        Visit visit = bahmniVisitService.getVisitSummary(visitUuid);
+        List<Encounter> encountersByVisitUuid = getEncounters(singletonList(visit));
+
+        Collection<Encounter> encountersByPatientProgramUuid = bahmniProgramWorkflowService
+                .getEncountersByPatientProgramUuid(patientProgramUuid);
+
+        if (isNotEmpty(encountersByVisitUuid) && isNotEmpty(encountersByPatientProgramUuid)) {
+            List<Encounter> encountersByPatientProgramUuidAndVisitUuid = encountersByPatientProgramUuid.stream()
+                    .filter(encounter -> encounter.getVisit().equals(visit)).collect(Collectors.toList());
+            return getFormDetails(patient, encountersByPatientProgramUuidAndVisitUuid, formType);
+        } else if (isNotEmpty(encountersByVisitUuid)) {
+            return getFormDetails(patient, encountersByVisitUuid, formType);
+        } else if (isNotEmpty(encountersByPatientProgramUuid)) {
+            return getFormDetails(patient, new ArrayList<>(encountersByPatientProgramUuid), formType);
+        }
+        return Collections.emptyList();
+    }
+
+    private List<Visit> limitVisits(List<Visit> visits, int numberOfVisits) {
+        if (numberOfVisits <= -1) {
```

Comment:
Should the check be against -1, or zero?

Reply:
It should be from -1. It gives an empty response for zero.

Reply:
It should be from -1. API gives an empty response if numberOfVisits are zero.

Reply:
Which API? BahmniFormDetailsController is invoking this service method passing the "numberOfVisits". The -1 value is because of @RequestParam(value = "numberOfVisits", defaultValue = "-1").